source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -535,7 +535,7 @@ void Filter::onUpstreamEvent(Network::ConnectionEvent event) {
     read_callbacks_->connection().readDisable(false);
 
     read_callbacks_->upstreamHost()->outlierDetector().putResult(
-        Upstream::Outlier::Result::SUCCESS);
+        Upstream::Outlier::Result::REQUEST_SUCCESS);
```

Comment:
@mattklein123 No - it should be REQUEST_SUCCESS. I had a long debugging session when testing consecutive 5xx in non-split mode. For some reason they did not kick-in and the reason was using REQUEST_SUCCESS vs CONNECT_SUCCESS. I will try to explain it here. Feel free to ask for more clarification if this is still not clear. This will also address your other comments why CONNECT_SUCCESS is handled differently and not mapped to HTTP 200 OK in non-split mode.

For router I added two events: CONNECT_FAILED and CONNECT_SUCCESS. CONNECT_FAILED usage is clear: it is called when Envoy detects localOrigin error. 

For CONNECT_SUCCESS, the sequence of events is as follows. Let us consider the case when connection to upstream server is OK, but the server repeatedly returns 5xx code. Because connection is OK, router will report two events to OutlierDetector: CONNECT_SUCCESS and actual received HTTP code 5xx. If CONNECT_SUCCESS was mapped to HTTP 200 OK, then for each request the outlier detector would receive sequence of {200, 5xx}, which is obviously wrong. In spit-mode CONNECT_SUCCESS goes to consecutive_local_origin bucket and does not intervene with HTTP codes returned from upstream server, but in non-split mode it must not be mapped to 200 OK.

For cases when you want the event to be mapped to HTTP code 200, REQUEST_SUCCESS should be used. And this is the case here in tcp_proxy: CONNECT_FAILED is mapped to 5xx error and REQUEST_SUCCESS is mapped to 200 OK. Therefore, in non-split mode the code behaves as before and configuration is backwards compatible.

Reply:
OK I see your point. I think what I'm saying is instead of having a single enum that tracks all errors, why not have 2 enums: `LocalOriginErrors` and `ExternalOriginErrors`. I think that will end up being a lot more clear through the code, regardless if split or not split is used?

Reply:
Let me think a bit about that. With 2 enum types there would be two interfaces to OutlierDetector. Let me go through the code and see how would it work.

Reply:
IMHO I do not think that having 2 APIs: one accepting LocalOriginErrors and second accepting ExternalOriginErrors would improve clarity much. Router reports HTTP codes through yet another API _putHTTPcode_ . The implementation of those APIs would be identical: map to HTTP code in non-split mode and call localOrigin methods for split mode. What do you think about renaming enum values we already have: instead of TIMEOUT, there would be LOCAL_ORIGIN_TIMEOUT, instead of REQUEST_SUCCESS there would be EXTERNAL_ORIGIN_REQUEST_SUCCESS, etc.

Reply:
I think prefixing with LOCAL_ORIGIN and EXTERNAL_ORIGIN would help, but I still don't understand why this is REQUEST_SUCCESS. This is the tcp_proxy. Why are we tracking "requests" here? Shouldn't this be just connection success and failure that are all local origin?

/wait-any

Reply:
I agree with you that intuitively it should be CONNECT_SUCCESS. But it goes again to my response to your comment titled: _Shouldn't this be CONNECT_SUCCESS?_. The problem is that in non-split mode CONNECT_SUCCESS is not mapped to any HTTP code. (If it was mapped to 200 OK, then router would receive two HTTP codes for each request). But for tcp_proxy, in non-split mode, we have to make a call to outlier detector with a parameter which is mapped to 200 OK. And this is what REQUEST_SUCCESS is doing.
But wait, maybe I can add optional parameter when reporting events to Outlier Detector. TIMEOUT is already doing it. Maybe I can always use CONNECT_SUCCESS and add 200 OK in tcp_proxy and NIL in router. It would be much clearer.

Reply:
+1. I think the other alternative here is to change the outlier detector APIs to make them work better for all the use cases we have now. Feel free to think about that. I think fundamentally this is problem is that it's too confusing to reason about what is going on with the names and APIs we have. Maybe approach it from that angle and see what falls out?

/wait-any

Reply:
I redesigned internal logic of reporting non-http results to outlier detector. In particular [putResult](https://github.com/cpakulski/envoy/blob/issue/3643/source/common/upstream/outlier_detection_impl.cc#L126-L139) method. LOCAL_ORIGIN_CONNECT_SUCCESS was previously handled in a specific way, which was hard to understand. Now, depending on the passed value of the _code_ there is no mapping to HTTP code, mapping based on defaults and mapping to passed HTTP code. The logic how to report events to outlier detector has been moved to individual filters, which in my opinion is much more clear.
As for API change to outlier detector, I am not sure whether much can be done without dropping some of the requirements, especially requirement to map non-http codes (LOCAL_ORIGIN_XXX and EXT_ORIGIN_XXX) to HTTP codes. Any sort of API would have to carry optional HTTP code, even when the code is later on not used in split mode.